gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java

```diff
@@ -306,16 +349,49 @@ private boolean shouldStopGeneratingWorkUnits(ConcurrentBoundedWorkUnitList work
   }
 
   @Override
-  public void shutdown(SourceState state) {}
+  public void shutdown(SourceState state) {
+  }
+
+  private Optional<CopyEntityWatermarkGenerator> getCopyEntityWatermarkGenerator(State state)
+      throws IOException {
+    try {
+      if (state.contains(WATERMARK_CREATOR)) {
+        Class<?> watermarkCreatorClass = Class.forName(state.getProp(WATERMARK_CREATOR));
+        return Optional.of((CopyEntityWatermarkGenerator) watermarkCreatorClass.newInstance());
+      } else {
+        return Optional.absent();
+      }
+    } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
+      throw new IOException("Failed to instantiate watermarkCreator.");
+    }
+  }
 
-  protected FileSystem getSourceFileSystem(State state) throws IOException {
+  private Optional<WatermarkInterval> getCopyEntityWatermark(CopyEntity copyEntity,
+      Optional<CopyEntityWatermarkGenerator> watermarkGenerator)
+      throws IOException {
+    if (!watermarkGenerator.isPresent()) {
+      return Optional.absent();
+    }
+    return watermarkGenerator.get().generateWatermarkIntervalForCopyEntity(copyEntity);
+  }
 
+  protected FileSystem getSourceFileSystem(State state)
+      throws IOException {
+    if (state.contains(CUSTOMIZED_FILESYSTEM)) {
```

Comment:
Usually this should be done by Hadoop itself. When you provide a fs URI, Hadoop will use the scheme to try to find a FileSystem matching that scheme. To provide custom `FileSystem` implementations you can either use the configuration "fs.<scheme>.impl", or register the implementation as a service provider for `FileSystem` (creating a file `META-INF/service/org.apache.hadoop.fs.FileSystem` containing the name of the implementation).